pkg/messageprocessors/javascript/javascript.go

```diff
@@ -40,113 +40,177 @@ func New() messageprocessors.PayloadEncodeDecoder {
 	}
 }
 
-func (h *host) createEnvironment(ids ttnpb.EndDeviceIdentifiers, version *ttnpb.EndDeviceVersionIdentifiers) map[string]interface{} {
-	env := make(map[string]interface{})
-	if ids.DevEUI != nil {
-		env["dev_eui"] = ids.DevEUI.String()
-	}
-	if version != nil {
-		env["brand"] = version.BrandID
-		env["model"] = version.ModelID
-		env["hardware_version"] = version.HardwareVersion
-		env["firmware_version"] = version.FirmwareVersion
-	}
-	return env
+type encodeDownlinkInput struct {
+	Data  map[string]interface{} `json:"data"`
+	FPort *uint8                 `json:"fPort"`
+}
+
+type encodeDownlinkOutput struct {
+	Bytes    []uint8  `json:"bytes"`
+	FPort    *uint8   `json:"fPort"`
+	Warnings []string `json:"warnings"`
```

Comment:
I see no usage for these warnings. Shouldn't we at least emit some events for these ?

Reply:
Yeah I was thinking what to do with this. Stay tuned.

Reply:
2f6888ec6

Reply:
```json
{
  "name": "as.up.data.decode.warning",
  "time": "2020-08-26T17:06:03.787144Z",
  "identifiers": [
    {
      "device_ids": {
        "device_id": "dev1",
        "application_ids": {
          "application_id": "app1"
        },
        "dev_eui": "1122334455667788",
        "join_eui": "0000000000000000"
      }
    }
  ],
  "data": {
    "@type": "type.googleapis.com/ttn.lorawan.v3.ApplicationUplink",
    "session_key_id": "AXQruu2BHxR68Zel+ZN1ag==",
    "f_port": 1,
    "f_cnt": 5,
    "frm_payload": "vu8=",
    "decoded_payload": {
        "bytes": [
              190,
              239
            ]
      },
    "decoded_payload_warnings": [
      "my test warning"
    ],
    "rx_metadata": [
      {
        "gateway_ids": {
          "gateway_id": "simulate"
        },
        "time": "2020-08-26T17:06:03.508552Z",
        "timestamp": 3584956232,
        "location": {
          "source": "SOURCE_REGISTRY"
        },
        "uplink_token": "CgwKCgoIc2ltdWxhdGUQyK64rQ0aDAj7rJr6BRC4npuMAiDA4quAq2g="
      }
    ],
    "settings": {
      "data_rate": {
        "lora": {
          "bandwidth": 125000,
          "spreading_factor": 12
        }
      },
      "coding_rate": "4/5",
      "frequency": "868500000",
      "timestamp": 3584956232,
      "time": "2020-08-26T17:06:03.508552Z"
    },
    "received_at": "2020-08-26T17:06:03.562813Z",
    "confirmed": true
  },
  "correlation_ids": [
    "as:up:01EGNVQMW5PN6F7YR6S3YPCA0F",
    "gs:conn:01EGNVQMMG00DZ6SQ6NS0PXJ3C",
    "gs:uplink:01EGNVQMNABPN41PK2HC4GDAH8",
    "ns:uplink:01EGNVQMNAC31GP3KSE6H8YA99",
    "rpc:/ttn.lorawan.v3.GsNs/HandleUplink:01EGNVQMNAC95K61P55C3S95KZ",
    "rpc:/ttn.lorawan.v3.GtwGs/LinkGateway:01EGNVQMKXN63FVC5TBZ05QCHJ"
  ],
  "origin": "Johan-2.local",
  "visibility": {
    "rights": [
      "RIGHT_APPLICATION_TRAFFIC_READ"
    ]
  }
}
```